File: Java/src/napakalaki/Player.java


package napakalaki;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;
import java.util.stream.Collectors;
import GUI.Dice;


public class Player {
    
    protected String name;
    protected int level = 1;
    protected boolean dead = true; 
    protected boolean canISteal = true;
    protected ArrayList<Treasure> hiddenTreasures = new ArrayList();
    protected ArrayList<Treasure> visibleTreasures = new ArrayList(); 
    protected BadConsequence pendingBC;
    protected Player enemy;
  
    
    
    public Player (String name){
        
        this.name = name;
    }
    
    public Player(Player p){
        
        this.name = p.name;
        this.level = p.level;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
        this.hiddenTreasures = p.hiddenTreasures;
        this.visibleTreasures = p.visibleTreasures;
        this.pendingBC = p.pendingBC;
        this.enemy = p.enemy;

    }
    
    public String getName() {
        
        return this.name;
    }
    
    private void bringToLife() {
        
        dead = false;
    }
    
    public boolean canISteal(){
        return canISteal && (this.pendingBC == null || this.pendingBC.isEmpty());
    }
    
    public int getCombatLevel() {

        return this.visibleTreasures.stream().map(Treasure::getBonus).reduce(this.level, Integer::sum);

    }
    
    public BadConsequence getPendingBC(){
        return this.pendingBC;
    }
    
    public Player getEnemy(){
        return enemy;
    }
    
    private void incrementLevels(int l){
        
        this.level += l;
    }
    
    private void decrementLevels(int l){
        
        this.level = this.level <= l ? 1 : this.level-l;

    }
    
    public boolean shouldConvert(){
        
        Dice dice = Dice.getInstance();
        return dice.nextNumber("Has perdido el combate.","En caso de sacar un 6, te convertiras en Sectario.") == 6;
        
    }
    
    private void setPendingBadconsequence (BadConsequence r){
        pendingBC = r;
    }
    
    private void applyPrize (Monster m) {
        
        this.incrementLevels(m.getPrize().getLevels());
        
        int nTreasures = m.getPrize().getTreasures();
            
        CardDealer cd = CardDealer.getInstance();
            
        for (int i = 0; i < nTreasures; i++) 
            this.hiddenTreasures.add(cd.nextTreasure());

        
    }
    
    private void applyBadConsequence (Monster m){
        
        this.decrementLevels(m.getBadconsequence().getLevels());
        
        this.setPendingBadconsequence(m.getBadconsequence().adjustToFitTreasureLists(this.visibleTreasures, this.hiddenTreasures));
              
    }
    
    private boolean canMakeTreasureVisible(Treasure t){
        
        if (t.getType() != TreasureKind.ONEHAND && t.getType() != TreasureKind.BOTHHANDS && !this.isTreasureKindInUse(t.getType())) return true;
        
        if (t.getType() == TreasureKind.ONEHAND && !this.isTreasureKindInUse(TreasureKind.BOTHHANDS)  && this.howManyVisibleTreasures(TreasureKind.ONEHAND) < 2) return true;
    
        return t.getType() == TreasureKind.BOTHHANDS && !this.isTreasureKindInUse(TreasureKind.BOTHHANDS)  && this.howManyVisibleTreasures(TreasureKind.ONEHAND) == 0;
    
    }
    
    protected Treasure giveMeATreasure(){
        
        return this.deleteAndReturnRandomIn(hiddenTreasures);
    }
    
    protected Treasure deleteAndReturnRandomIn(ArrayList<Treasure> list){
        
        Random rd = new Random();
        Treasure t = list.get(rd.nextInt() % list.size());
        
        list.remove(t);
        
        this.dieIfNoTreasures();
        
        return t;
    }
    
    private int howManyVisibleTreasures(TreasureKind tk){

        int b = Collections.frequency(this.visibleTreasures.stream().map(a -> a.getType()).collect(Collectors.toList()), tk);
        return b;
    }
    
    private boolean isTreasureKindInUse(TreasureKind type) {
        
        boolean c = this.visibleTreasures.stream().map( a -> a.getType() ).collect(Collectors.toList()).contains(type);
        return c;
    }
    
    private void dieIfNoTreasures() {
        
        if(this.visibleTreasures.isEmpty() && this.hiddenTreasures.isEmpty()) 
            this.dead = true;    
    }
    
    public boolean isDead() {
        
        return this.dead;
    }

    public ArrayList<Treasure> getHiddenTreasures() {
        return hiddenTreasures;
    }

    public ArrayList<Treasure> getVisibleTreasures() {
        return visibleTreasures;
    }
    
    protected int getOponentLevel(Monster m){
        
        return m.getCombatLevel();
    }
    
    public CombatResult combat(Monster m) {
        
        Dice dice = Dice.getInstance();
        
        int lv = getOponentLevel(m) + (!canISteal() && dice.nextNumber("Determinará si se suma el nivel del enemigo.", "Esto ocurrirá en caso de sacar 1 o 2.") < 3 ? this.enemy.getCombatLevel() : 0);
        
        if (this.getCombatLevel() <= lv){
            
            this.applyBadConsequence(m);
            
            if (shouldConvert()) return CombatResult.LOSEANDCONVERT;
            
            return CombatResult.LOSE;
        }
        
        
        this.applyPrize(m);
        
        if(this.level >= 10) return CombatResult.WINGAME;
        
        return CombatResult.WIN;
        
    }
    
    public void makeTreasureVisible(Treasure t){
        
        if(this.canMakeTreasureVisible(t)) {
            
            this.visibleTreasures.add(t);
            
            this.hiddenTreasures.remove(t);
            
        }
    }

    public void discardHiddenTreasure(Treasure t){
        
        CardDealer cd = CardDealer.getInstance();
        
        cd.giveTreasureBack(t);

        this.hiddenTreasures.remove(t);
        
        if (this.pendingBC != null && !this.pendingBC.isEmpty()) {

            this.pendingBC.substractHiddenTreasure(t);

        }

        this.dieIfNoTreasures();
    }
    
     public void discardVisibleTreasure(Treasure t){

        CardDealer cd = CardDealer.getInstance();
        
        cd.giveTreasureBack(t);
                
        this.visibleTreasures.remove(t);
        
        if (this.pendingBC != null && !this.pendingBC.isEmpty()) {

            this.pendingBC.substractVisibleTreasure(t);

        }
        
        this.dieIfNoTreasures();
    
    }
    
    public boolean ValidState(){
        
        return (this.pendingBC == null || this.pendingBC.isEmpty()) && this.hiddenTreasures.size() <= 4;
        
    }
    
    public void initTreasures(){
        
        CardDealer dealer = CardDealer.getInstance();
        
        Dice dice = Dice.getInstance();
        
        this.bringToLife();
        
        Treasure treasure;
        
        int number = dice.nextNumber("INICIALIZACIÓN DE TESOROS"," ");
        
        
        treasure = dealer.nextTreasure();
        this.hiddenTreasures.add(treasure);   
        
        if (number == 6) {
            
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
        if (number > 1){
            treasure = dealer.nextTreasure();
            this.hiddenTreasures.add(treasure);
        }
      
    }
    
    
    public int getLevels() {
        
        return this.level;
    }
    
    public Treasure stealTreasure(){
        
        if(!isAbleToSteal()) return null;
        
        Treasure t = this.enemy.giveMeATreasure();
        this.hiddenTreasures.add(t);
        
        this.haveStolen();
        
        return t;
    }   
    
    protected boolean canYouGiveMeATreasure(){
        return !this.hiddenTreasures.isEmpty();
    }
    
    public boolean isAbleToSteal(){
        return this.canISteal && this.enemy.canYouGiveMeATreasure();
    }
    
    public void haveStolen(){
        this.canISteal = false;
    }
    
    public void discardAllTreasures(){
        
        ArrayList<Treasure> visTreasures = new ArrayList(visibleTreasures);
        ArrayList<Treasure> hidTreasures = new ArrayList(hiddenTreasures);

        for(Treasure t : visTreasures){
            this.discardVisibleTreasure(t);
        }
        
        for(Treasure t : hidTreasures){
            this.discardHiddenTreasure(t);
        }
    }
    
    public void setEnemy (Player enemy){
        
        this.enemy = enemy;
    
    }
    
    @Override
    public String toString() {
        return name + " Nivel del jugador: " + level + " Nivel de combate: " + getCombatLevel() + " Enemigo: " + enemy.getName();
    }


}

File: Ruby/napakalaki/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame

require_relative 'bad_consequence.rb'
require_relative "specific_bad_consequence.rb"
require_relative "numeric_bad_consequence.rb"
require_relative "death_bad_consequence.rb"
require_relative 'prize.rb'
require_relative 'monster.rb'
require_relative 'treasure.rb'
require_relative 'combat_result.rb'
require_relative 'card_dealer.rb'
require_relative 'dice.rb'

class Player
  
  def initialize(name)
    @name = name
    @visibleTreasures = Array.new
    @hiddenTreasures = Array.new
    @death = true
    @level = 1    
    @canISteal = true;
    
  end
  
  
  def copy(player)
    
    @name = player.getName
    @visibleTreasures = player.getVisibleTreasures
    @hiddenTreasures = player.getHiddenTreasures
    @death = player.isDead
    @canISteal = player.canISteal
    @enemy = player.getEnemy
    @level = player.getLevel
    @pendingBC = player.getBC
    
  end
  
  def bringToLife
    
    @death = false
    
  end
  
  def getName
    @name
  end
  
  def getLevel
    @level
  end
  
  def getBC
    @pendingBC
  end
  
  def getEnemy
    @enemy
  end
  
  def isDead
    @death
  end
  
  def shouldConvert
    
    dice = Dice.instance
    n = dice.nextNumber
    return n == 6
    
  end
  
  def getVisibleTreasures
    @visibleTreasures
  end
  
  def getHiddenTreasures
    @hiddenTreasures
  end
  
  def getCombatLevel
   
    @visibleTreasures.inject(@level) { |acc,t| acc + t.bonus }

  end
  
  def incrementLevels(n)
    
    @level += n
    
  end
  
  def decrementLevels(n)
    
    @level = @level <= n ? 1 : @level - n
    
  end
   
  def setPendingBC(bc)
    
    @pendingBC = bc
  
  end
  
  
  def applyPrize(m)
    
    incrementLevels(m.getPrize.levels)
    
    n = m.getPrize.treasures
    
    cd = CardDealer.instance
    
    n.times do
      @hiddenTreasures << cd.nextTreasure
    end
   
  end
  
  def applyBadConsequence(m)
    
    decrementLevels(m.getBadConsequence.nLevels)
 
    setPendingBC(m.getBadConsequence.adjustToFitTreasureList(@visibleTreasures, @hiddenTreasures))
    
  end
  
  def canMakeTreasureVisible(t)
    
    if(t.type != TreasureKind::ONEHAND && t.type != TreasureKind::BOTHHANDS && !isTreasureKindInUse(t.type)) 
      return true
    end
    
    if t.type == TreasureKind::ONEHAND && !isTreasureKindInUse(TreasureKind::BOTHHANDS) && howManyVisibleTreasures(t.type) < 2
      return true
    end
    
    return t.type == TreasureKind::BOTHHANDS && !isTreasureKindInUse(t.type) && howManyVisibleTreasures(TreasureKind::ONEHAND) == 0

  end
  
  def canYouGiveMeATreasure
    @hiddenTreasures.any?
  end
  
  def haveStolen
    @canISteal = false
  end
  
  def canISteal
    @canISteal
  end
  
  def discardAllTreasures
    
    @visibleTreasures.each do |t|
      discardVisibleTreasure(t)
    end
    
    @hiddenTreasures.each do |t|
      discardHiddenTreasure(t)
    end
  end
  
  
  def howManyVisibleTreasures(tk)
    
    
    @visibleTreasures.collect{ |a| a.type}.count(tk)

  end
  
  
  def isTreasureKindInUse(type)
        
    @visibleTreasures.collect{ |a| a.type}.include?(type)

  end
  
  def getOponentLevel(m)
    m.getCombatLevel
  end
  
  def dieIfNoTreasures
    
    @death = true if @visibleTreasures.empty? && @hiddenTreasures.empty?
  end
  
  def combat (m)
    
    dice = Dice.instance
    n = dice.nextNumber
    
    level = getOponentLevel(m) +  (!canISteal && n < 3 ? @enemy.getCombatLevel : 0)
    
    if getCombatLevel <= level then
      
      applyBadConsequence(m)
      
      return CombatResult::LOSEANDCONVERT if shouldConvert
      
      return CombatResult::LOSE
      
    end
    
    applyPrize(m)
    
    if @level >= 10 
      return CombatResult::WINGAME
    end
    
    return CombatResult::WIN
    
  end
  
  
  def makeTreasureVisible(t)
    
    if canMakeTreasureVisible(t) then
      
      @visibleTreasures << @hiddenTreasures.delete(t)

    end
  end
  
  def discardVisibleTreasure(t)
    
    @visibleTreasures -= [t]
    
    c = CardDealer.instance
    
    c.giveTreasureBack(t);
    
    if @pendingBC != nil && !@pendingBC.empty? then
      
      @pendingBC.substractVisibleTreasure(t)
    
    end
    
    dieIfNoTreasures
    
  end

  def discardHiddenTreasure(t)
    
    @hiddenTreasures -= [t]
    
    c = CardDealer.instance
    
    c.giveTreasureBack(t);
    
    if @pendingBC != nil && !@pendingBC.empty? then
      
      @pendingBC.substractHiddenTreasure(t)
    
    end
    
    dieIfNoTreasures
    
  end
  
  def validState
    
    return @hiddenTreasures.length <= 4 && (@pendingBC.nil? || @pendingBC.empty?)
    
  end
  
  def stealTreasure
    
    unless @canISteal && @enemy.canYouGiveMeATreasure
      return nil
    end
        
    haveStolen
    
    t = @enemy.giveMeATreasure
    @hiddenTreasures << t

   return t
  end
  
  def giveMeATreasure
    
    t = @hiddenTreasures[rand(@hiddenTreasures.size)]
    discardHiddenTreasure(t)
    return t
    
  end
  
  def initTreasures
    
    dealer = CardDealer.instance
    dice = Dice.instance
    
    bringToLife
    
    treasure = dealer.nextTreasure
    @hiddenTreasures << treasure
    
    n = dice.nextNumber
    
    if n == 6 then
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      
    elsif n > 1 then
      
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      
    end

  end
  
  def setEnemy (enemy)
    @enemy = enemy
  end
  
  def to_s
    "#{@name} (Nivel de jugador: #{@level}  Nivel de combate: #{getCombatLevel}  Enemigo: #{@enemy.getName})"
  end
  
  private :bringToLife, :incrementLevels, :decrementLevels, :setPendingBC, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures
  protected :giveMeATreasure, :canYouGiveMeATreasure, :getCombatLevel
  
end
end
